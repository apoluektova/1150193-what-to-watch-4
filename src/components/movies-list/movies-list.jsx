import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import MovieCard from "../movie-card/movie-card.jsx";

class MoviesList extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      activeMovieCard: null,
    };
  }

  render() {
    const {movies, onMovieCardClick} = this.props;

    return (
      <div className="catalog__movies-list">
        {movies.map((movie, index) => {
          return (
            <MovieCard
              key={index}
              movie={movie}
              onMovieCardClick={onMovieCardClick}
              onMovieCardHover={(currentMovie) => {
                this.setState({
                  activeMovieCard: currentMovie,
                });
              }}
            />
          );
        })}
      </div>
    );
  }
}

MoviesList.propTypes = {
  movies: PropTypes.array.isRequired,
  onMovieCardClick: PropTypes.func.isRequired,
};

export default MoviesList;
