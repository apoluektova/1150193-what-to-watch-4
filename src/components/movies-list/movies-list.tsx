import * as React from "react";
import MovieCard from "../movie-card/movie-card";
import withVideoPlayer from "../../hocs/with-video-player/with-video-player";
import {connect} from "react-redux";
import ShowMoreButton from "../show-more-button/show-more-button";
import {ActionCreator} from "../../reducer/app/app";
import {getFilteredMovies} from "../../reducer/data/selectors";
import {getShownMovieCards} from "../../reducer/app/selectors";
import {Movie} from "../../types";

interface Props {
  movies: Array<Movie>;
  onMovieCardClick: () => void;
  onActiveCardChange: () => void;
  onShowMoreButtonClick: () => void;
  shownMovieCards: number;
}

const MovieCardWrapped = withVideoPlayer(MovieCard);

const MoviesList: React.FunctionComponent<Props> = (props: Props) => {
  const {
    movies,
    onMovieCardClick,
    onActiveCardChange,
    onShowMoreButtonClick,
    shownMovieCards
  } = props;

  const shownMovies = movies.slice(0, shownMovieCards);

  return (
    <React.Fragment>
      <div className="catalog__movies-list">
        {shownMovies.map((movie) => {
          return (
            <MovieCardWrapped
              key={movie.id}
              movie={movie}
              onMovieCardClick={onMovieCardClick}
              onMovieCardHover={onActiveCardChange}
            />
          );
        })}
      </div>
      <div className="catalog__more">
        {shownMovieCards < movies.length && <ShowMoreButton
          onShowMoreButtonClick={onShowMoreButtonClick}
        />}
      </div>
    </React.Fragment>
  );
};

const mapStateToProps = (state) => ({
  movies: getFilteredMovies(state),
  shownMovieCards: getShownMovieCards(state),
});

const mapDispatchToProps = (dispatch) => ({
  onShowMoreButtonClick() {
    dispatch(ActionCreator.showMoreMovies());
  }
});

export {MoviesList};
export default connect(mapStateToProps, mapDispatchToProps)(MoviesList);


