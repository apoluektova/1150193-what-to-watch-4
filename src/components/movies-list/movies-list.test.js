import React from "react";
import renderer from "react-test-renderer";
import {MoviesList} from "./movies-list.jsx";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from '../../reducer/name-space.js';

const mockStore = configureStore([]);

const movies = [
  {
    id: 1,
    previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    videoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    title: `Fantastic Beasts: The Crimes of Grindelwald`,
    backgroundImage: `https://loremflickr.com/cache/resized/65535_49876816733_f1ba86707f_h_1280_543_nofilter.jpg`,
    backgroundColor: `#A6B7AC`,
    poster: `https://loremflickr.com/cache/resized/65535_49824720108_18907b74af_z_273_410_nofilter.jpg`,
    genre: `Fantasy`,
    releaseDate: 2018,
    description: `The plot follows Newt Scamander and Albus Dumbledore as they attempt to take down the dark wizard Gellert Grindelwald while facing new threats in a more divided wizarding world.`,
    rating: {
      score: 8.9,
      count: 240,
    },
    director: `David Yates`,
    actors: [`Michael Fassbender`, `Marion Cotillard`, `Paddy Considine`, `Sean Harris`],
    runtime: `1h 39m`,
    isFavorite: false,
  },
  {
    id: 2,
    previewImage: `img/bohemian-rhapsody.jpg`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    title: `Bohemian Rhapsody`,
    backgroundImage: `https://loremflickr.com/cache/resized/65535_49768198796_957c97bc00_h_1280_543_nofilter.jpg`,
    backgroundColor: `#A6B7AC`,
    poster: `https://loremflickr.com/cache/resized/65535_50001660108_922f0950ea_z_273_410_nofilter.jpg`,
    genre: `Drama`,
    releaseDate: 2018,
    description: `A British-American venture, the film was produced by 20th Century Fox, Regency Enterprises, GK Films, and Queen Films, with Fox serving as distributor. The film follows the singer's life from the formation of the band up to their 1985 Live Aid performance at the original Wembley Stadium.`,
    rating: {
      score: 9.0,
      count: 250,
    },
    director: `Bryan Singer`,
    actors: [`Michael Fassbender`, `Marion Cotillard`, `Paddy Considine`, `Sean Harris`],
    runtime: `1h 39m`,
    isFavorite: false,
  },
  {
    id: 3,
    previewImage: `img/macbeth.jpg`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    title: `Macbeth`,
    backgroundImage: `https://loremflickr.com/cache/resized/65535_49757902562_3b737630e4_h_1280_543_nofilter.jpg`,
    backgroundColor: `#A6B7AC`,
    poster: `https://loremflickr.com/cache/resized/65535_49816709832_c9df27e040_z_273_410_nofilter.jpg`,
    genre: `Drama`,
    releaseDate: 2015,
    description: `The story follows a Scottish general's rise to power after receiving a prophecy from a trio of witches that one day he will become King of Scotland. Like the play it was adapted from, the film dramatises the damaging physical and psychological effects of political ambition on those who seek power for its own sake.`,
    rating: {
      score: 8.0,
      count: 234,
    },
    director: `Justin Kurzel`,
    actors: [`Michael Fassbender`, `Marion Cotillard`, `Paddy Considine`, `Sean Harris`],
    runtime: `1h 39m`,
    isFavorite: false,
  },
];

it(`MoviesList should render correctly`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      movies,
    },
    [NameSpace.APP]: {
      shownMovieCards: 8,
    },
  });

  const tree = renderer
     .create(
         <Provider store={store}>
           <MoviesList
             movies={movies}
             onMovieCardClick={() => {}}
             onActiveCardChange={() => {}}
             handleShowMoreButtonClick={() => {}}
             shownMovieCards={8}
           />
         </Provider>, {
           createNodeMock: () => {
             return {};
           }
         })
     .toJSON();

  expect(tree).toMatchSnapshot();
});
