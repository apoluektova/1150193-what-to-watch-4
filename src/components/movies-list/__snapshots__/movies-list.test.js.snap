// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList should render correctly 1`] = `
Array [
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <video
          className="player__video"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="movie-page.html"
          onClick={[Function]}
        >
          Fantastic Beasts: The Crimes of Grindelwald
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <video
          className="player__video"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="movie-page.html"
          onClick={[Function]}
        >
          Bohemian Rhapsody
        </a>
      </h3>
    </article>
  </div>,
  <div
    className="catalog__more"
  />,
]
`;
